server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java

```diff
@@ -214,4 +291,170 @@ private void initLookup(SortedSetDocValues dvs) throws IOException {
+            return DocValues.getSortedSet(context.reader(), field);
+        }
+
+        /**
+         * For the first time, this iterator is allowed to skip documents. It needs to pull {@link
+         * PostingsEnum}s from the terms dictionary of the inverted index and create a priority queue
+         * out of them.
+         */
+        private void init(long minOrd, long maxOrd) throws IOException {
+            dynamicPruningInitialized++;
+
+            final int size = (int) Math.max(0, maxOrd - minOrd + 1);
+            postings = new ArrayDeque<>(size);
+            if (size > 0) {
+                Terms terms = context.reader().terms(field);
+                if (terms != null) {
+                    BytesRef minTerm = BytesRef.deepCopyOf(lookup.lookupOrd(minOrd));
+                    TermsEnum termsEnum = terms.iterator();
+                    TermsEnum.SeekStatus seekStatus = termsEnum.seekCeil(minTerm);
+                    if (seekStatus != TermsEnum.SeekStatus.END) {
+                        BytesRef maxTerm = BytesRef.deepCopyOf(lookup.lookupOrd(maxOrd));
+
+                        TermsEnum globalTermsEnum = lookup.termsEnum();
+                        globalTermsEnum.seekExact(minOrd);
+
+                        for (BytesRef term = termsEnum.term(); term != null && term.compareTo(maxTerm) <= 0; term = termsEnum.next()) {
+                            // Compute the global ordinal of this term by advancing the global terms enum to the same term and retrieving
+                            // the term ordinal. This is cheaper than calling lookupTerm for every term.
+                            while (globalTermsEnum.term().compareTo(term) < 0) {
+                                BytesRef nextGlobalTerm = globalTermsEnum.next();
+                                assert nextGlobalTerm != null;
+                            }
+                            assert globalTermsEnum.term().equals(term);
+                            final long globalOrd = globalTermsEnum.ord();
+                            postings.add(new PostingsEnumAndOrd(termsEnum.postings(null, PostingsEnum.NONE), globalOrd));
+                        }
+                    }
+                }
+            }
+            disjunction = new PriorityQueue<>(size) {
+                @Override
+                protected boolean lessThan(PostingsEnumAndOrd a, PostingsEnumAndOrd b) {
+                    return a.postings.docID() < b.postings.docID();
+                }
+            };
+            disjunction.addAll(postings);
+        }
+
+        public void updateBounds() throws IOException {
+            if (highestCompetitiveValueGlobalOrd == null) {
+                return;
+            }
+
+            long lowOrd;
+            if (afterValueGlobalOrd != null && afterValueGlobalOrd != MISSING_VALUE_FLAG) {
+                lowOrd = afterValueGlobalOrd;
```

Comment:
When there is a single source, we should be able to bump this by 1 I think, since the after key is excluded? The implementation might be a bit tricky and require updating initLookup to record the first ordinal that compares strictly greater than `afterKey`, which might be the same as `lowOrd` when the `afterKey` does not exist in the index. We could look into it in a follow-up PR and just leave a comment for now.

Reply:
Bumping by 1 when there is a single source makes sense to me. This should be beneficial on its own. I do not fully understand the benefit of also recording the first ordinal > `afterKey`. Even though the competitive iterator might have a wider range on the low side, the actual query should skip non-matching docs anyway, right?

Reply:
There is a corner case when the after key does not exist in a shard. For instance say a shard has two values `a` and `c`, and the `after_key` is `b`. The `after` global ord will be set to the global ordinal of `c` in that case, so it shouldn't be bumped here.

Reply:
I would expect us to have a check on whether there is a single source (we can't skip the ord of the after key if there are other sources, e.g. say your after key is `[ "c", "d" ]`, then `["c", "e"]` also compares greater than the after key, while the leading source has a value equal to the same value as in the after key). But indeed, something like that. Let's make sure we have good tests for this case to make sure we don't skip more than would be correct?

Reply:
Right yes. OK, I think I prefer to add a TODO about this additional optimization for now and leave it for a follow up PR.

Reply:
I will raise an issue as soon as this is merged.

Reply:
I raised this in https://github.com/elastic/elasticsearch/issues/92505